Client/Frontend/Brave Notifications/BraveNotificationsPresenter.swift

```diff
@@ -0,0 +1,489 @@
+  }
+  
+  private var panState: CGPoint = .zero
+  @objc private func dismissPannedAdView(_ pan: UIPanGestureRecognizer) {
+    guard let notificationView = pan.view else { return }
+    
+    switch pan.state {
+    case .began:
+      panState = notificationView.center
+      // Make sure to stop the dismiss timer
+      dismissTimers[notificationView]?.invalidate()
+    case .changed:
+      notificationView.transform.ty = min(0, pan.translation(in: notificationView).y)
+    case .ended:
+      let velocity = pan.velocity(in: notificationView).y
+      let y = min(panState.y, panState.y + pan.translation(in: notificationView).y)
+      let projected = project(initialVelocity: velocity, decelerationRate: UIScrollView.DecelerationRate.normal.rawValue)
+      if y + projected < 0 {
+        guard let notification = self.displayedNotifications.first(where: { $0.view == notificationView }) else { return }
+        hide(notificationView: notificationView, velocity: velocity)
+        notification.willDismiss(timedout: false)
+        break
+      }
+      fallthrough
+    case .cancelled:
+      // Re-setup timeout timer
+      guard let notification = self.displayedNotifications.first(where: { $0.view == notificationView }) else { return }
+      if case .automatic(let interval) = notification.dismissPolicy {
+        setupTimeoutTimer(for: notification, interval: interval)
+      }
+      notificationView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+        animation.toValue = 0
+      }
+    default:
+      break
+    }
+  }
+  
+  // MARK: - Animations
+  
+  private func animateIn(adView: UIView) {
+    adView.layoutIfNeeded()
+    adView.layer.transform = CATransform3DMakeTranslation(0, -adView.bounds.size.height, 0)
+    
+    adView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+      animation.toValue = 0
+    }
+  }
+  
+  private func animateOut(adView: UIView, velocity: CGFloat? = nil, completion: @escaping () -> Void) {
+    adView.layoutIfNeeded()
+    let y = adView.frame.minY - view.safeAreaInsets.top - adView.transform.ty
+    
+    adView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+      animation.toValue = -(view.safeAreaInsets.top + y + adView.bounds.size.height)
+      if let velocity = velocity {
+        animation.velocity = velocity
+      }
+      animation.completionBlock = { _, _ in
+        completion()
+      }
+    }
+  }
+}
+
+extension BraveNotificationsPresenter {
+  class View: UIView {
+    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
+      // Only allow tapping the ad part of this VC
+      if let view = super.hitTest(point, with: event), view.superview is AdView {
+        return view
+      } else if let view = super.hitTest(point, with: event), view is WalletConnectionView {
+        return view
+      }
```

Comment:
This shouldn't look for specific view's, compare it with the visible notification's view